i12974 - two new docs:... (#737)
codenvy/che-docs: src/main/pages/che-7/hands-on-guidance/proc_configuring-a-che-workspace-using-a-devfile.adoc

[id="configuring-a-che-workspace-using-a-devfile_{context}"]
= Configuring a Che workspace using a Devfile

This section will show how to select a workspace, loading in the required stack and a sample project, all in one step.

Please refer to the link:using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces[Devfile user guide] and link:https://redhat-developer.github.io/devfile/devfile[Devfile specifications] for detailed information how to add or remove components, or how to set memory and environment variables.

> 

Reviewer: ```suggestion
This section describes how to select a workspace and load the required stack and a sample project. For detailed information on adding or removing components, or to set memory and environment variables, see link:using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces[devfile user guide] and link:https://redhat-developer.github.io/devfile/devfile[devfile specifications].
```

Reviewer 2: @supriya-bharadwaj, there's no such thing as 'devfile user guide'. Please, link to a specific section.

Author: the link is to using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces ... is that not a devfile user guide?

Author: Also in my version it was Devfile, not devfile. Can we agree it should be capitalized?

Reviewer 2: @nickboldt,

>  is that not a devfile user guide?

No, it's a section/chapter in a larger doc. "User guide" is ambiguous. We try to name (and then link to) sections as specifically as possible. In this case, we're linking to the "Making a workspace portable using a devfile" section.

> Also in my version it was Devfile, not devfile. Can we agree it should be capitalized?

No, it shouldn't be capitalized (unless in the beginning of a sentence). Devfile is not a proper noun. Same with workspace, factory, plug-in, stack, etc.

@supriya-bharadwaj, sections don't walk (us)... :) Don't anthropomorphize. Use "This section describes how to..."

Reviewer: @rkratky , point noted. :)
(I am unlearning so much)

[discrete]
== Prerequisites

* You need an OKD, OCP, or OSD cluster, if using Che on Openshift. See below for use of link:https://www.okd.io/minishift/[Minishift].
* Or, you need a Kubernetes cluster such as link:https://github.com/kubernetes/minikube#installation[Minikube] to use Che on Kubernetes.

> 

Reviewer: ```suggestion
* For Che on OpenShift, an OKD, OCP, or OSD cluster. To use Minishift, see https://www.okd.io/minishift/. 
* For Che on Kubernetes, a Kubernetes cluster like link:https://github.com/kubernetes/minikube#installation[Minikube].
```

[discrete]
== Procedure

=== Importing a project and workspace in one step using a devfile

> 

Author: we don't want to emphasize that with a devfile, it's ONE step instead of 2+?

Reviewer 2: @nickboldt, your orig. headline suggests there are also other ways to do it using a devfile (in two steps, in three steps) :)

Generally, docs are not the place to offer judgement on the quality of the described procedures. If we want to emphasize that something is easier with a devfile than it was before, then the release notes are the place for it.

In Che 7, stacks are defined as link:using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces[devfiles], instead of container images. This affords more flexibility when defining the project(s), IDE, commands, tools, and application runtimes needed in your workspace.

> 

Author: you have "the user" and "your". Pick just one person - 3rd or 2nd, but not both. :)

Reviewer 2: +1

(@supriya-bharadwaj, pick the 'user'. Docs should never address the reader directly. Not only does it sound bad, but it's also incorrect because the reader may not be the person who's actually doing the stuff -- unless you're providing instructions in the form of "do this, do that".)

Reviewer: Sure, @rkratky .

So where in the past you might have included a `Dockerfile` in your project's repo, now you can include a `devfile.yaml` file in the root of your git project instead.

> 

Author: The point here is that the devfile is a replacement for the need for a dockerfile.

Reviewer 2: @nickboldt, the instructions should not include references to was before, what isn't yet implemented, etc. That said, if there's a need to call attention to differences from what was used in the past, then we should use an admonition (instructions say "do this, do that", and then we add "Note that in Che 6, blah blah, but in Che 7, blah blah").

@supriya-bharadwaj, don't say "you can" in docs. The user either should do it, or should not. If the user has multiple options and needs to decide, then describe what each option does.

==== The minimal devfile

> The minimum content you need in a `devfile.yaml` file is:

Reviewer: ```suggestion
The following is the minimum content required in a `devfile.yaml` file:
```

* link:https://redhat-developer.github.io/devfile/devfile#apiversion[apiVersion]
* link:https://redhat-developer.github.io/devfile/devfile#metadata[metadata name]
* link:https://redhat-developer.github.io/devfile/devfile#projects[projects name and source]

```
apiVersion: 1.0.0
metadata:
  name: che-in-che-out
projects:
  - name: che
    source:
      type: git
      location: 'https://github.com/eclipse/che.git'
```

> For a more complete example, see link:https://github.com/eclipse/che/blob/master/devfile.yaml[Eclipse Che in Che devfile.yaml].

Reviewer: ```suggestion
For a complete example of a `devfile.yaml` file, see https://github.com/eclipse/che/blob/master/devfile.yaml[Eclipse Che in Che devfile.yaml].
```

Reviewer 2: @supriya-bharadwaj, the orig. link was correct. You introduced a naked URL, which we try to avoid.

Reviewer: Edited.

==== Creating a workspace from a devfile

> 

Reviewer: ```suggestion
To create a workspace from a devfile:
```

> . Commit a `devfile.yaml` into the root of your git repo.

Reviewer: ```suggestion
. Commit a `devfile.yaml` to the root of your Git repository. 
```

Reviewer 2: @supriya-bharadwaj, "into" was correct. The user is not committing "in" the root directory. The root directory is the destination, so "into" is fine ("to" would be even better, but not "in").

Reviewer: Rectified.

. Once your repo contains a `devfile.yaml`, point a running Che 7 instance at that link:https://github.com/eclipse/che[project] and it will use the discovered link:https://github.com/eclipse/che/blob/master/devfile.yaml[devfile.yaml] to build you a workspace via its Factory API, `/f?url=`.

> 

Reviewer 2: @supriya-bharadwaj, you have left out the API.

Reviewer: Added it back in.

Author: not sure what needs to be changed here, @rkratky @supriya-bharadwaj

Author: I've added a link to the forthcoming Factory API page, eg. at  http://0.0.0.0:4000/che-7/api-reference.html#factory-api_api-reference

[id="configuring-a-che-workspace-using-a-devfile_{context}"]
= Configuring a Che workspace using a Devfile

This section will show how to select a workspace, loading in the required stack and a sample project, all in one step.

Please refer to the link:using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces[Devfile user guide] and link:https://redhat-developer.github.io/devfile/devfile[Devfile specifications] for detailed information how to add or remove components, or how to set memory and environment variables.

[discrete]
== Prerequisites

* You need an OKD, OCP, or OSD cluster, if using Che on Openshift. See below for use of link:https://www.okd.io/minishift/[Minishift].
* Or, you need a Kubernetes cluster such as link:https://github.com/kubernetes/minikube#installation[Minikube] to use Che on Kubernetes.
* For Che on OpenShift, an OKD, OCP, or OSD cluster. To use Minishift, see https://www.okd.io/minishift/. 
* For Che on Kubernetes, a Kubernetes cluster like link:https://github.com/kubernetes/minikube#installation[Minikube].
[discrete]
== Procedure

> === Importing a project and workspace in one step using a devfile

Reviewer: ```suggestion
=== Importing a project and workspace using a devfile
```

> In Che 7, stacks are defined as link:using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces[devfiles], instead of container images. This affords more flexibility when defining the project(s), IDE, commands, tools, and application runtimes needed in your workspace.

Reviewer: ```suggestion
In Che 7, stacks are defined as link:using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces[devfiles]. This gives the user flexibility when defining the project(s), IDE, commands, tools, and application runtimes needed in the workspace.
```

Author: You reviewed the outdated copy. New copy reads:

```
Stacks are defined as link:using-developer-environments-workspaces.html#making-a-workspace-portable-using-a-devfile_using-developer-environments-workspaces[devfiles]. This gives the user more flexibility when defining the project(s), IDE, commands, tools, and application runtimes needed in their workspace.
```

> So where in the past you might have included a `Dockerfile` in your project's repo, now you can include a `devfile.yaml` file in the root of your git project instead.

Reviewer: ```suggestion
In Che 7, the user must include a `devfile.yaml` file at the root of the Git project. 
```

Reviewer: @nickboldt , does this make sense?

==== The minimal devfile

The minimum content you need in a `devfile.yaml` file is:

* link:https://redhat-developer.github.io/devfile/devfile#apiversion[apiVersion]
* link:https://redhat-developer.github.io/devfile/devfile#metadata[metadata name]
* link:https://redhat-developer.github.io/devfile/devfile#projects[projects name and source]

```
apiVersion: 1.0.0
metadata:
  name: che-in-che-out
projects:
  - name: che
    source:
      type: git
      location: 'https://github.com/eclipse/che.git'
```

For a more complete example, see link:https://github.com/eclipse/che/blob/master/devfile.yaml[Eclipse Che in Che devfile.yaml].

==== Creating a workspace from a devfile

. Commit a `devfile.yaml` into the root of your git repo.

> . Once your repo contains a `devfile.yaml`, point a running Che 7 instance at that link:https://github.com/eclipse/che[project] and it will use the discovered link:https://github.com/eclipse/che/blob/master/devfile.yaml[devfile.yaml] to build you a workspace via its Factory API, `/f?url=`.

Reviewer: ```suggestion
. When the repository contains the `devfile.yaml` file, point a running Che instance to the project at link:https://github.com/eclipse/che[]. Che will use the discovered `devfile.yaml` file (from link:https://github.com/eclipse/che/blob/master/devfile.yaml[]) and build a workspace using the factory API, `/f?url=`.
```

Author: You made that way more confusing. 
New copy:

```
. Commit a `devfile.yaml` file into the root of your git repo.

. When your repository contains a `devfile.yaml` file, point a running Che 7 instance at that link:https://github.com/eclipse/che[project] and it will use the discovered link:https://github.com/eclipse/che/blob/master/devfile.yaml[devfile.yaml] file to build a workspace using the xref:api-reference#factory-api_api-reference[Factory API], `/f?url=`.

* For example, if you're using xref:che-quick-starts#deploying-che-using-chectl[Che inside Minishift]:

```

If you're going to remove the "your" references in one of the bullets, you should also remove it from the others... but as you didn't change "your git repo" or "if you're using..." I figure we should be consistent and keep all three in 2nd person voice.
